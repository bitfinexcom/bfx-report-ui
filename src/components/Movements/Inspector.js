import React from 'react'
import PropTypes from 'prop-types'

const INSPECTORS = {
  ETH: {
    name: 'etherscan.io',
    url: 'https://etherscan.io/address/',
  },
}

export const Inspector = ({ currency, destinationAddress }) => {
  const target = INSPECTORS[currency]
  return target ? (
    <span className='bitfinex-show-soft'>
      (
      <a href={`${target.url}${destinationAddress}`} target='_blank' rel='noopener noreferrer'>
        {target.name}
      </a>
)
    </span>
  ) : ''
}

Inspector.propTypes = {
  currency: PropTypes.string,
  destinationAddress: PropTypes.string,
}
Inspector.defaultProps = {
  currency: '',
  destinationAddress: '',
}

export default Inspector
