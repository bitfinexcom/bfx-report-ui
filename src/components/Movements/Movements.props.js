import PropTypes from 'prop-types'
import { intlShape } from 'react-intl'

const MOVEMENTS_ENTRIES_PROPS = PropTypes.shape({
  id: PropTypes.number.isRequired,
  currency: PropTypes.string.isRequired,
  mtsStarted: PropTypes.number.isRequired,
  mtsUpdated: PropTypes.number.isRequired,
  status: PropTypes.string.isRequired,
  amount: PropTypes.number.isRequired,
  destinationAddress: PropTypes.string,
})

export const propTypes = {
  entries: PropTypes.arrayOf(MOVEMENTS_ENTRIES_PROPS).isRequired,
  intl: intlShape.isRequired,
  loading: PropTypes.bool.isRequired,
}

export const defaultProps = {
  entries: [],
  intl: {},
  loading: true,
}
