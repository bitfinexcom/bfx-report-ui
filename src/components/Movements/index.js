export { default } from './Movements.container'
