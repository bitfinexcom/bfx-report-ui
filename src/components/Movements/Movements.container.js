import { connect } from 'react-redux'
import Movements from './Movements'

function mapStateToProps(state = {}) {
  return {
    entries: state.movements.entries,
    loading: !state.movements.dataReceived,
  }
}

const MovementsContainer = connect(mapStateToProps)(Movements)

export default MovementsContainer
