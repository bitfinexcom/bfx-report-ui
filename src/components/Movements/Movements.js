import React, { Fragment, PureComponent } from 'react'
import { injectIntl } from 'react-intl'
import {
  Button,
  Card,
  Elevation,
} from '@blueprintjs/core'
import {
  Cell,
  Column,
  Table,
  TruncatedFormat,
} from '@blueprintjs/table'
import Loading from 'components/Loading'
import NoData from 'components/NoData'
import Pagination from 'components/Pagination'
import queryConstants from 'state/query/constants'
import { checkFetch, formatTime, getCurrentEntries } from 'state/utils'
import { propTypes, defaultProps } from './Movements.props'
import Inspector from './Inspector'

// const TYPE_DEPOSITS = 'deposits'
const TYPE_WITHDRAWALS = 'withdrawals'
const COLUMN_WIDTHS = [80, 150, 125, 120, 400]
const LIMIT = queryConstants.DEFAULT_MOVEMENTS_QUERY_LIMIT
const PAGE_SIZE = queryConstants.DEFAULT_MOVEMENTS_PAGE_SIZE

class Movements extends PureComponent {
  constructor(props) {
    super(props)
    this.fetchPrev = this.fetchPrev.bind(this)
    this.fetchNext = this.fetchNext.bind(this)
  }

  componentDidMount() {
    // eslint-disable-next-line react/destructuring-assignment
    this.props.fetchMovements()
  }

  componentDidUpdate(prevProps) {
    checkFetch(prevProps, this.props, 'movements')
  }

  fetchPrev() {
    // eslint-disable-next-line react/destructuring-assignment
    this.props.fetchPrevMovements()
  }

  fetchNext() {
    // eslint-disable-next-line react/destructuring-assignment
    this.props.fetchNextMovements()
  }

  render() {
    const {
      offset,
      pageOffset,
      entries,
      intl,
      jumpPage,
      type,
      loading,
    } = this.props
    const currentEntries = getCurrentEntries(entries, offset, LIMIT, pageOffset, PAGE_SIZE)
    const filteredData = currentEntries.filter(entry => (type === TYPE_WITHDRAWALS
      ? parseFloat(entry.amount) < 0 : parseFloat(entry.amount) > 0))
    const numRows = filteredData.length

    const idCellRenderer = rowIndex => (
      <Cell>
        {filteredData[rowIndex].id}
      </Cell>
    )

    const mtsUpdatedCellRenderer = rowIndex => (
      <Cell>
        <TruncatedFormat>
          {formatTime(filteredData[rowIndex].mtsUpdated)}
        </TruncatedFormat>
      </Cell>
    )

    const amountCellRenderer = rowIndex => (
      <Cell className='bitfinex-text-align-right'>
        {filteredData[rowIndex].amount}
        &nbsp;
        <span className='bitfinex-show-soft'>
          {filteredData[rowIndex].currency}
        </span>
      </Cell>
    )

    const statusCellRenderer = rowIndex => (
      <Cell>
        {filteredData[rowIndex].status}
      </Cell>
    )

    const destinationCellRenderer = (rowIndex) => {
      const entry = filteredData[rowIndex]
      return (
        <Cell>
          {entry.destinationAddress}
          &nbsp;
          <Inspector currency={entry.currency} destinationAddress={entry.destinationAddress} />
        </Cell>
      )
    }

    const titleMsgId = type === TYPE_WITHDRAWALS
      ? 'movements.withdrawals.title' : 'movements.deposits.title'
    let showContent
    if (loading) {
      showContent = (
        <Loading title={titleMsgId} />
      )
    } else if (numRows === 0) {
      showContent = (
        <NoData title={titleMsgId} />
      )
    } else {
      showContent = (
        <Fragment>
          <h4>
            {intl.formatMessage({ id: titleMsgId })}
            &nbsp;
            <Button icon='cloud-download' disabled>
              {intl.formatMessage({ id: 'timeframe.download' })}
            </Button>
          </h4>
          <Table
            className='bitfinex-table'
            numRows={numRows}
            enableRowHeader={false}
            columnWidths={COLUMN_WIDTHS}
          >
            <Column
              id='id'
              name='#'
              cellRenderer={idCellRenderer}
            />
            <Column
              id='mtsupdated'
              name={intl.formatMessage({ id: 'movements.column.updated' })}
              cellRenderer={mtsUpdatedCellRenderer}
            />
            <Column
              id='status'
              name={intl.formatMessage({ id: 'movements.column.status' })}
              cellRenderer={statusCellRenderer}
            />
            <Column
              id='amount'
              name={intl.formatMessage({ id: 'movements.column.amount' })}
              cellRenderer={amountCellRenderer}
            />
            <Column
              id='destination'
              name={intl.formatMessage({ id: 'movements.column.destination' })}
              cellRenderer={destinationCellRenderer}
            />
          </Table>
          <Pagination
            type='movements'
            dataLen={entries.length}
            offset={offset}
            jumpPage={jumpPage}
            prevClick={this.fetchPrev}
            nextClick={this.fetchNext}
            pageOffset={pageOffset}
          />
        </Fragment>
      )
    }
    return (
      <Card interactive elevation={Elevation.ZERO} className='col-lg-12 col-md-12 col-sm-12 col-xs-12'>
        {showContent}
      </Card>
    )
  }
}

Movements.propTypes = propTypes
Movements.defaultProps = defaultProps

export default injectIntl(Movements)
