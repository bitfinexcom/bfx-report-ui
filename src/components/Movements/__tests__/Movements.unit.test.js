import React from 'react'
import ReactDOM from 'react-dom'
import {
  renderWithIntl,
  rendererCreateWithIntl,
  shallowWithIntl,
} from 'tests/helper'
import Movements from '../Movements'

const NO_ENTRY = []
const ENTRIES = [
  {
    id: 10450678,
    currency: 'USD',
    currencyName: 'WIRE',
    mtsStarted: 1530801307000,
    mtsUpdated: 1530804653000,
    status: 'COMPLETED',
    amount: 1234,
    fees: 0,
    destinationAddress: null,
    transactionId: null,
  },
  {
    id: 10450289,
    currency: 'ETH',
    currencyName: 'ETHEREUM',
    mtsStarted: 1527418430000,
    mtsUpdated: 1527418789000,
    status: 'COMPLETED',
    amount: 12.3,
    fees: 0,
    destinationAddress: '0x1234c77b33444514babcded17ee0f6197f0e3ff4',
    transactionId: '0xb62d0ca12fad7294ec1234c611233bea66e4b37d0e4a101bf7a4dfd06cf8261c',
  },
  {
    id: 1350463,
    currency: 'ETH',
    currencyName: 'ETHEREUM',
    mtsStarted: 1528251777000,
    mtsUpdated: 1528251998000,
    status: 'PENDING REVIEW',
    amount: -0.9976,
    fees: -0.0024,
    destinationAddress: '0x12345678c3E5833A1664fA7F8A757D226669f3c6',
    transactionId: null,
  },
]

test('deposits renders without crashing', () => {
  const div = document.createElement('div')
  renderWithIntl(<Movements type='deposits' />, div)
  ReactDOM.unmountComponentAtNode(div)
})

test('withdrawals renders without crashing', () => {
  const div = document.createElement('div')
  renderWithIntl(<Movements type='withdrawals' />, div)
  ReactDOM.unmountComponentAtNode(div)
})

test('deposits renders without crashing with data', () => {
  shallowWithIntl(<Movements type='deposits' loading={false} entries={ENTRIES} />)
})

test('withdrawals renders without crashing with data', () => {
  shallowWithIntl(<Movements type='withdrawals' loading={false} entries={ENTRIES} />)
})

test('deposits render loading correctly', () => {
  const component = rendererCreateWithIntl(<Movements type='deposits' />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('withdrawals render loading correctly', () => {
  const component = rendererCreateWithIntl(<Movements type='withdrawals' />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('deposits render no data correctly', () => {
  const component = rendererCreateWithIntl(<Movements type='deposits' loading={false} entries={NO_ENTRY} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('withdrawals render no data correctly', () => {
  const component = rendererCreateWithIntl(<Movements type='withdrawals' loading={false} entries={NO_ENTRY} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

// test('deposits render correctly with data', () => {
//   const component = rendererCreateWithIntl(<Movements type='deposits' loading={false} entries={ENTRIES} />)
//   const tree = component.toJSON()
//   expect(tree).toMatchSnapshot()
// })

// test('withdrawals render correctly with data', () => {
//   const component = rendererCreateWithIntl(<Movements type='withdrawals' loading={false} entries={ENTRIES} />)
//   const tree = component.toJSON()
//   expect(tree).toMatchSnapshot()
// })
