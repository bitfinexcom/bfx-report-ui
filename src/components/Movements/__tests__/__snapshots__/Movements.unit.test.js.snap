// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deposits render loading correctly 1`] = `
<div
  className="bp3-card bp3-interactive bp3-elevation-0 col-lg-12 col-md-12 col-sm-12 col-xs-12"
>
  <h4>
    Deposits
  </h4>
  <svg
    className="bp3-spinner"
    height={24}
    strokeWidth={16}
    viewBox="0 0 100 100"
    width={24}
  >
    <path
      className="bp3-spinner-track"
      d="M 50,50 m 0,-44.5 a 44.5,44.5 0 1 1 0,89 a 44.5,44.5 0 1 1 0,-89"
    />
    <path
      className="bp3-spinner-head"
      d="M 50,50 m 0,-44.5 a 44.5,44.5 0 1 1 0,89 a 44.5,44.5 0 1 1 0,-89"
      pathLength={280}
      strokeDasharray="280 280"
      strokeDashoffset={210}
    />
  </svg>
</div>
`;

exports[`deposits render no data correctly 1`] = `
<div
  className="bp3-card bp3-interactive bp3-elevation-0 col-lg-12 col-md-12 col-sm-12 col-xs-12"
>
  <h4>
    Deposits
  </h4>
  <div>
    No related data in this time range. You can try another time range.
  </div>
</div>
`;

exports[`withdrawals render loading correctly 1`] = `
<div
  className="bp3-card bp3-interactive bp3-elevation-0 col-lg-12 col-md-12 col-sm-12 col-xs-12"
>
  <h4>
    Withdrawals
  </h4>
  <svg
    className="bp3-spinner"
    height={24}
    strokeWidth={16}
    viewBox="0 0 100 100"
    width={24}
  >
    <path
      className="bp3-spinner-track"
      d="M 50,50 m 0,-44.5 a 44.5,44.5 0 1 1 0,89 a 44.5,44.5 0 1 1 0,-89"
    />
    <path
      className="bp3-spinner-head"
      d="M 50,50 m 0,-44.5 a 44.5,44.5 0 1 1 0,89 a 44.5,44.5 0 1 1 0,-89"
      pathLength={280}
      strokeDasharray="280 280"
      strokeDashoffset={210}
    />
  </svg>
</div>
`;

exports[`withdrawals render no data correctly 1`] = `
<div
  className="bp3-card bp3-interactive bp3-elevation-0 col-lg-12 col-md-12 col-sm-12 col-xs-12"
>
  <h4>
    Withdrawals
  </h4>
  <div>
    No related data in this time range. You can try another time range.
  </div>
</div>
`;
